Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean

```diff
@@ -0,0 +1,287 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Affine.Basic
+
+/-!
+
+# Singular cubics
+
+Let `E` be a cubic over an arbitrary field `K`,
+given by a Weierstrass equation `y² + a₁xy + a₃y = x³ + a₂x² + a₄x + a₆`.
+
+We define `E.singular` as a typeclass containing the data of a singular point.
+(Such a point is unique if exists.
+See `WeierstrassCurve.IsSingularPoint.ext` or `WeierstrassCurve.Singular.subsingleton`.)
+
+For curves `E` with `[E.Singular]`, we define `E.SplitSingular` to contain the data of
+(one of) the tangents at the singular point.
+
+## TODO (Andrew)
+- Classify singular cubics
+
+-/
+
+namespace WeierstrassCurve
+
+variable {R : Type*} [CommRing R] (E : WeierstrassCurve R)
+
+/-- The predicate of being a singular point on an `WeierstrassCurve` in affine coordinates. -/
+def Affine.IsSingularPoint (E : WeierstrassCurve.Affine R) (x y : R) : Prop :=
+  E.Equation x y ∧ E.polynomialX.evalEval x y = 0 ∧ E.polynomialY.evalEval x y = 0
+
+lemma Affine.isSingularPoint_iff {E : WeierstrassCurve.Affine R} {x y : R} :
+    E.IsSingularPoint x y ↔ E.Equation x y ∧
+      E.a₁ * y - (3 * x ^ 2 + 2 * E.a₂ * x + E.a₄) = 0 ∧ 2 * y + E.a₁ * x + E.a₃ = 0 := by
+  rw [IsSingularPoint, evalEval_polynomialX, evalEval_polynomialY]
+
+lemma Affine.IsSingularPoint.relation_x
+    {E : WeierstrassCurve.Affine R} {x y : R} (h : E.IsSingularPoint x y) :
+    E.a₁ * y = 3 * x ^ 2 + 2 * E.a₂ * x + E.a₄ :=
+  sub_eq_zero.mp ((Affine.isSingularPoint_iff.mp h).2.1)
+
+lemma Affine.IsSingularPoint.relation_y
+    {E : WeierstrassCurve.Affine R} {x y : R} (h : E.IsSingularPoint x y) :
+    2 * y + E.a₁ * x + E.a₃ = 0 :=
+  (Affine.isSingularPoint_iff.mp h).2.2
```

Comment:
Do you need these? I had `Affine.nonsingular_iff'` as the analogue of your `Affine.isSingularPoint_iff` and `Affine.nonsingular_iff` for a cleaner version without the brackets like these two lemmas.

Reply:
This is for dot notation.